dot/parachain/statement-distribution/received_manifests.go

```diff
@@ -0,0 +1,182 @@
+//
+// This assumes that the manifest has already been checked for
+// validity - i.e. that the bitvecs match the claimed group in size
+// and that the manifest includes at least one `Seconded`
+// attestation and includes enough attestations for the candidate
+// to be backed.
+//
+// This also should only be invoked when we are intended to track
+// the knowledge of this peer as determined by the [SessionTopology].
+func (rm *receivedManifests) importReceived(
+	groupSize uint,
+	secondingLimit uint,
+	candidateHash parachaintypes.CandidateHash,
+	manifestSummary manifestSummary,
+) error {
+	previousSummary, ok := rm.received[candidateHash]
+	if !ok {
+		withinLimits := updatingEnsureWithinSecondingLimit(
+			rm.secondedCounts,
+			manifestSummary.claimedGroupIndex,
+			groupSize,
+			secondingLimit,
+			manifestSummary.statementKnowledge.secondedInGroup,
+		)
+
+		if withinLimits {
+			rm.received[candidateHash] = manifestSummary
+			return nil
+		} else {
+			return errManifestImportOverflow
+		}
+	}
+
+	if previousSummary.claimedGroupIndex != manifestSummary.claimedGroupIndex {
+		return errManifestImportConflicting
+	}
+
+	if !manifestSummary.statementKnowledge.secondedInGroup.Contains(
+		previousSummary.statementKnowledge.secondedInGroup,
+	) {
+		return errManifestImportConflicting
+	}
+
+	if !manifestSummary.statementKnowledge.validatedInGroup.Contains(
+		previousSummary.statementKnowledge.validatedInGroup,
+	) {
+		return errManifestImportConflicting
+	}
+
+	freshSeconded, err := manifestSummary.statementKnowledge.secondedInGroup.Or(
+		previousSummary.statementKnowledge.secondedInGroup,
+	)
+	if err != nil {
+		return err
+	}
+
+	withinLimits := updatingEnsureWithinSecondingLimit(
+		rm.secondedCounts,
+		manifestSummary.claimedGroupIndex,
+		groupSize,
+		secondingLimit,
+		freshSeconded,
+	)
+
+	if !withinLimits {
+		return errManifestImportOverflow
+	}
+
+	// All checks passed. Overwrite: guaranteed to be superset.
+	rm.received[candidateHash] = manifestSummary
+	return nil
+}
+
+// updatingEnsureWithinSecondingLimit updates validator-seconded records but only
+// if the new statements are OK. returns `true` if alright and `false` otherwise.
+//
+// The seconding limit is a per-validator limit. It ensures an upper bound on the total number of
+// candidates entering the system.
+//
+// The function mutates secondedCounts.
+func updatingEnsureWithinSecondingLimit(
+	secondedCounts map[parachaintypes.GroupIndex][]uint,
+	groupIndex parachaintypes.GroupIndex,
+	groupSize uint,
+	secondingLimit uint,
+	newSeconded parachaintypes.BitVec, // polkadot-sdk uses a &BitSlice<u8, Lsb0>
+) bool {
+	if secondingLimit == 0 {
+		return false
+	}
+
+	counts, ok := secondedCounts[groupIndex]
+	if !ok {
+		counts = make([]uint, groupSize)
+		secondedCounts[groupIndex] = counts
+	}
+
+	nsBits := newSeconded.Bits()
+	for i, bit := range nsBits {
+		if !bit {
+			continue
+		}
+
+		if i < len(counts)-1 && counts[i] == secondingLimit {
```

Comment:
`i < len(counts)` is enough given that when `i == len(counts)` it will be out of bounds
```suggestion
		if i < len(counts) && counts[i] == secondingLimit {
```

Reply:
Good catch, thanks!

I've modified the test case `existing_counts_would_exceed_limit` to cover  this corner case.